### DISCUSSION TITLE
Does anyone know how you can synchronize the playback of several esp32?

### DISCUSSION THREAD

**Discussion by franciscogimeno2000** [baldram/ESP_VS1053_Library#88] [Q&A]
Hello.. and thanks for your time..
The tests with this library work fine, but I'm trying to synchronize the playback on several esp32s so that it sounds the same at the same time and I don't know where to start... does anyone have any ideas?

Greetings and thank you
Fran

**Comment by baldram** [comment]
Good question! Kind of multi-room solution. I'm also very interested in ideas on synchronization.

**Reply by franciscogimeno2000** [reply]
I have read that the WMA and Ogg Vorbis protocol sends the real time of the current position of the player in the encoding. But I don't know how to read this.

This pdf is interesting https://sylvan.fish/docs/synplay.pdf

**Comment by Dr-Dawg** [comment]
Sounds great!! :-)

My ideas wold be:

- use some sort of ring buffer, like the modification of @fabitom :
https://github.com/fabitom/ESP_VS1053_Library/blob/ring-buffer-demo-v2/examples/WebRadioDemo/RingBufferDemo.cpp
- there are four sync bytes in the mp3 stream. If you scan for them and manage to play the same part of the buffer at the the same time index on all ESPs, this would do the trick..
- maybe using streamModeOn() and  streamModeOff() might help to synchronize

Please let us know about any progress!

**Reply by Dr-Dawg** [reply]
Instead of analyzing the buffer, using the feedback described in the following part of theVS1053b Datasheet might also work:
 9.6.9 SCI_HDAT0 and SCI_HDAT1 (R)
This would of course cause additions to the existing library :-)

**Comment by franciscogimeno2000** [comment]
Hello friends, after much looking... I have verified that equipment such as the arylic S10, what they do to create multiroom through the api or the app:
From a group of devices you tell each of them which is going to be the master player for that area and the others take the synchronization data from that selected master and not from the streaming player.

**Comment by franciscogimeno2000** [comment]
In this way you can create a simpler form of synchronization and as you want, since it does not matter to you how and what type the sound source and the server type are.
The synchronization data will be taken from the master unit, for example by a selected port against the fixed IP of the master unit programmed via tcp or udp.

**Reply by Dr-Dawg** [reply]
From my experience, putting up an ESP32 WebServer that is working parallel to the client that is reading the stream is quite easy.
In my case the server acts as a webinterface to control the radio, set the station parameters and choose MIDI instruments. I'm using ESPAsyncWebServer.h as described here:
https://randomnerdtutorials.com/esp32-web-server-spiffs-spi-flash-file-system/
However, I'm not sure whether the ESPs capabilities are sufficient to provide several clients with streaming data without any latencies, the supposed buffer and synchronizing the clients (like you described) in any form might still be necessary. Also, I think the master device will provide the data to the clients on their request, I assume this will require to separate the buffer that is filled in the main loop from the data that is provided to the clients.

Once this is accomplished it would be a small step to add a mp3 (file) Server reading from a sdcard and providing your client players with your favourite music :-)
